rfcs/0000-top-level-assumptions.md

```diff
@@ -0,0 +1,327 @@
+
+# How we teach this
+
+This RFC proposed the addition of a new set of options in a way that every option can affect multiple plugins, and a plugin can be affected by multiple options.
+
+Documenting the different `loose` and `loose`-like options wasn't too hard: every plugin had a bunch of those options, and we could document them in the plugin's page.
+
+A possible structure for the documentation could be a two-entry table, with all the plugins in the rows and the assumptions in the columns, marking which assumption affects which plugins. However, this would lead to an enormous highly sparse table, impossible to fit in a normal web page.
+
+I think that the best way of documenting these options is to put all of them in a single long list, divided in two sections: assumptions that affect standard features, and assumptions that only affect proposals. For each assumption we should write:
+- The option name
+- It's description
+- An example (which could be interactive, using CodeSandbox)
+- The affected plugins
+
+In the docs of each affected plugin, we should write the assumptions that affect it and link to the corresponding entry in the assumptions list. We shouldn't duplicate the description of what each assumption does, because duplicating it would likely lead to the two copies become out of sync.
+
+We can also provide a tool to check that the all the assumptions hold. This can't be done with an ESLint plugin because almost all the assumptions are about runtime semantics hard to statically analyze, but we can do it using a Babel plugin which should run while testing, and which injects assertions in the transpiled code:
+
+```javascript 
+// input code
+
+foo?.bar;
+```
+
+```javascript
+// output code
+
+(() => {
+  const _tmp = foo;
+  if (_tmp == null && _tmp !== null && _tmp !== undefined) {
+    throw new Error(`
+      Invalid assumption - noDocumentAll.
+      
+      "noDocumentAll: true" implies that ?. and ?? are never
+      used with the special document.all object.
+      This assumption has been violated, so you should
+      disable it in your config.
+    `);
+  }
+  return _tmp;
+})?.bar
+```
+
+<details>
+<summary>Config</summary>
+
+```jsonc
+// babel.config.json
+{
+  "assumptions": {
+    "noDocumentAll": true
+  },
+  "env": {
+    "test": {
+      "plugins": ["@babel/validate-assumptions"]
+    }
+  }
+}
+```
+
+</details>
+
+# Open questions
+
+- Should we try to pass to the presets at least a _partial_ `assumptions` object? Currently `@babel/preset-env` relies on `loose` to enable/disable the `typeof-symbol` plugin.
+- Should we validate the list of `assumptions` in `@babel/core`, and disallow unknown ones? This would make it impossible for third-party plugins to introduce their own assumptions, but it also means that it's easier for us to introduce new assumptions without risking ecosystem incompatibilities.
```

Comment:
https://github.com/babel/rfcs/pull/5#issuecomment-641547262 (@existentialism)
> 2. My gut/initial reaction is that we should validate the list of `assumptions`

---

https://github.com/babel/rfcs/pull/5#issuecomment-643668256 (@nicolo-ribaudo)

> I initially wasn't 100% on board with it because it means that if we add support for a new assumptions in a plugin, then the user would need to update `@babel/core` in order to use it (otherwise Babel would throw an "unknown assumption" error).
However, I think that the benefits of validating outweigh this downside :+1:

---

I was going to write in the RFC text that `@babel/core` will throw on unknown assumptions, but I realized that if we add a new assumptions in `@babel/core@7.15`, presets cannot use it because it would be a breaking change when used with `@babel/core@7.14`.

Note that this problem will likely never affect our own presets (since we won't enable assumptions in our own presets), but it could affect framework-specific ones.

We have a few options:
1. Decide that it's ok, presets can check Babel's version and conditionally enable an assumption.
1. Only throw for unknown assumption in config files, not in presets.
1. Throw in config files, warn in presets.
1. Always warn.
1. Don't validate the assumptions list.

In any case, we shouldn't pass unknown assumptions down to plugins.

Reply:
Agreed. Note that `@babel/preset-env` also validate unknown options. If a preset depends on `@babel/preset-env`, every time when a new option is introduced, they have to bump dependencies, too. I think it applies to `assumptions` of `@babel/core`, too.